packages/block-editor/src/components/block-title/use-block-display-title.js

```diff
@@ -0,0 +1,83 @@
+/**
+ * External dependencies
+ */
+import { truncate } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { useSelect } from '@wordpress/data';
+import {
+	getBlockType,
+	__experimentalGetBlockLabel as getBlockLabel,
+	isReusableBlock,
+} from '@wordpress/blocks';
+
+/**
+ * Internal dependencies
+ */
+import useBlockDisplayInformation from '../use-block-display-information';
+import { store as blockEditorStore } from '../../store';
+
+/**
+ * Returns the block's configured title as a string, or empty if the title
+ * cannot be determined.
+ *
+ * @example
+ *
+ * ```js
+ * useBlockDisplayTitle( 'afd1cb17-2c08-4e7a-91be-007ba7ddc3a1', 17 );
+ * ```
+ *
+ * @param {string} clientId Client ID of block.
+ * @param {number|undefined} maximumLength The maximum length that the block title string may be before truncated.
+ * @return {?string} Block title.
+ */
+export default function useBlockDisplayTitle( clientId, maximumLength ) {
+	const { attributes, name, reusableBlockTitle } = useSelect(
+		( select ) => {
+			if ( ! clientId ) {
+				return {};
+			}
+			const {
+				getBlockName,
+				getBlockAttributes,
+				__experimentalGetReusableBlockTitle,
+			} = select( blockEditorStore );
+			const blockName = getBlockName( clientId );
+			if ( ! blockName ) {
+				return {};
+			}
+			const isReusable = isReusableBlock( getBlockType( blockName ) );
+			return {
+				attributes: getBlockAttributes( clientId ),
+				name: blockName,
+				reusableBlockTitle:
+					isReusable &&
+					__experimentalGetReusableBlockTitle(
+						getBlockAttributes( clientId ).ref
+					),
+			};
+		},
+		[ clientId ]
+	);
+
+	const blockInformation = useBlockDisplayInformation( clientId );
+	if ( ! name || ! blockInformation ) {
+		return null;
+	}
+	const blockType = getBlockType( name );
+	const blockLabel = blockType
+		? getBlockLabel( blockType, attributes )
+		: null;
+	const label = reusableBlockTitle || blockLabel;
+	// Label will fallback to the title if no label is defined for the current
+	// label context. If the label is defined we prioritize it over possible
+	// possible block variation title match.
+	if ( label && label !== blockType.title ) {
+		return maximumLength && maximumLength > 0
+			? truncate( label, { length: maximumLength } )
+			: label;
+	}
```

Comment:
It's minor, but it might be worth mentioning it in the docs for the `maximumLength` param that truncation only applies to the `label`.

Reply:
Good point. Definitely worth communicating that. I'll throw up a small PR.

Reply:
https://github.com/WordPress/gutenberg/pull/39328

Reply:
What if the block title has 100 characters? 😄 

Maybe we should also truncate it for consistency when someone explicitly passes this parameter.

Reply:
> What if the block title has 100 characters? 😄

💥 

Haha. Yeah, the expectation seems to have been that block title strings should be protected from truncation or were never going to be long enough to be truncated.

> Maybe we should also truncate it for consistency when someone explicitly passes this parameter.

A cursory grep through the `block.json` files (there might be a few e2e fixtures in there) shows that there's not much over 28 chars in length, so we could probably get away with it.

<details><summary>Block title string lengths</summary>

```bash
Read More: 12
Site Logo: 12
Post Author: 14
Video: 8
Post Comments Link: 21
Comment Author Avatar: 24
Custom Link: 14
Comments Pagination: 22
Tag Cloud: 12
Calendar: 11
Buttons: 10
Post Date: 12
Comments Query Loop: 22
Embed: 8
Post Excerpt: 15
File: 7
Paragraph: 12
Quote: 8
Group: 8
Site Tagline: 15
Shortcode: 12
Comment Edit Link: 20
Separator: 12
Gallery: 10
Heading: 10
Comment Template: 19
Comment Content: 18
Term Description: 19
RSS: 6
Post Author Name: 19
Next Page: 12
Page Numbers: 15
Code: 7
Navigation: 13
Verse: 8
Post Comments Count: 22
Pullquote: 12
Post Author Biography: 24
Query Title: 14
Custom HTML: 14
Column: 9
Template Part: 16
Archives: 11
Social Icons: 15
Unsupported: 14
Spacer: 9
Social Icon: 14
Post Navigation Link: 23
Image: 8
More: 7
Table of Contents: 20
Search: 9
Page Break: 13
Text Columns (deprecated): 28
Comment Date: 15
Audio: 8
Post Comments Form: 21
Next Page: 12
Button: 9
Login/out: 12
Table: 8
Post Template: 16
List: 7
Preformatted: 15
Pagination: 13
Classic: 10
Post Terms: 13
Previous Page: 16
Navigation Area: 18
Home Link: 12
Post Featured Image: 22
Page List: 12
Previous Page: 16
Comment Author Name: 22
Columns: 10
Latest Comments: 18
Latest Posts: 15
Pattern: 10
Media & Text: 15
Post Content: 15
Post Title: 13
Query Loop: 13
Site Title: 13
Comment Reply Link: 21
Submenu: 10
Page Numbers: 15
Categories: 13
Post Comments: 16
Reusable block: 17
Post Comment (deprecated): 28
Cover: 8
Iframed Block: 16
Iframed Masonry Block: 24
Iframed Multiple Stylesheets: 31
Iframed Inline Styles: 24
Legacy Widget: 16
Notice: 9
Read More: 12
Site Logo: 12
Post Author: 14
Video: 8
Post Comments Link: 21
Comment Author Avatar: 24
Custom Link: 14
Comments Pagination: 22
Tag Cloud: 12
Calendar: 11
Buttons: 10
Post Date: 12
Comments Query Loop: 22
Embed: 8
Post Excerpt: 15
File: 7
Paragraph: 12
Quote: 8
Group: 8
Site Tagline: 15
Shortcode: 12
Comment Edit Link: 20
Separator: 12
Gallery: 10
Heading: 10
Comment Template: 19
Comment Content: 18
Term Description: 19
RSS: 6
Post Author Name: 19
Next Page: 12
Page Numbers: 15
Code: 7
Navigation: 13
Verse: 8
Post Comments Count: 22
Pullquote: 12
Post Author Biography: 24
Query Title: 14
Custom HTML: 14
Column: 9
Template Part: 16
Archives: 11
Social Icons: 15
Unsupported: 14
Spacer: 9
Social Icon: 14
Post Navigation Link: 23
Image: 8
More: 7
Table of Contents: 20
Search: 9
Page Break: 13
Text Columns (deprecated): 28
Comment Date: 15
Audio: 8
Post Comments Form: 21
Next Page: 12
Button: 9
Login/out: 12
Table: 8
Post Template: 16
List: 7
Preformatted: 15
Pagination: 13
Classic: 10
Post Terms: 13
Previous Page: 16
Navigation Area: 18
Home Link: 12
Post Featured Image: 22
Page List: 12
Previous Page: 16
Comment Author Name: 22
Columns: 10
Latest Comments: 18
Latest Posts: 15
Pattern: 10
Media & Text: 15
Post Content: 15
Post Title: 13
Query Loop: 13
Site Title: 13
Comment Reply Link: 21
Submenu: 10
Page Numbers: 15
Categories: 13
Post Comments: 16
Reusable block: 17
Post Comment (deprecated): 28
Cover: 8
Iframed Block: 16
Iframed Masonry Block: 24
Iframed Multiple Stylesheets: 31
Iframed Inline Styles: 24
Legacy Widget: 16
Read More: 12
Post Comment Author: 22
Site Logo: 12
Post Author: 14
Post Comment Date: 20
Video: 8
Post Comments Link: 21
Comment Author Avatar: 24
Comment Author Avatar: 24
Custom Link: 14
Comments Pagination: 22
Tag Cloud: 12
Calendar: 11
Buttons: 10
Post Date: 12
Comments Query Loop: 22
Embed: 8
Post Comment Content: 23
Post Excerpt: 15
File: 7
Paragraph: 12
Quote: 8
Group: 8
Site Tagline: 15
Shortcode: 12
Post Comment Reply Link: 26
Comment Edit Link: 20
Separator: 12
Gallery: 10
Heading: 10
Comment Template: 19
Comment Content: 18
Term Description: 19
RSS: 6
Post Author Name: 19
Next Page: 12
Page Numbers: 15
Code: 7
Navigation: 13
Verse: 8
Post Comments Count: 22
Pullquote: 12
Post Author Biography: 24
Query Title: 14
Custom HTML: 14
Column: 9
Template Part: 16
Archives: 11
Social Icons: 15
Unsupported: 14
Spacer: 9
Social Icon: 14
Post Navigation Link: 23
Image: 8
More: 7
Table of Contents: 20
Search: 9
Page Break: 13
Text Columns (deprecated): 28
Comment Date: 15
Audio: 8
Post Comments Form: 21
Next Page: 12
Button: 9
Login/out: 12
Table: 8
Post Template: 16
List: 7
Preformatted: 15
Pagination: 13
Classic: 10
Post Terms: 13
Previous Page: 16
Navigation Area: 18
Home Link: 12
Post Featured Image: 22
Page List: 12
Previous Page: 16
Comment Author Name: 22
Post Empty: 13
Post Comment Edit Link: 25
Columns: 10
Latest Comments: 18
Latest Posts: 15
Pattern: 10
Media & Text: 15
Post Content: 15
Post Title: 13
Query Loop: 13
Site Title: 13
Comment Reply Link: 21
Submenu: 10
Page Numbers: 15
Categories: 13
Post Comments: 16
Reusable block: 17
Post Comment (deprecated): 28
Cover: 8
Legacy Widget: 16
Notice: 9
```
</details>

Reply:
If the truncation is applied to both the block title and the other labels, that's probably the point at which it doesn't need to be part of the hook.

The same code would just be:
```js
const title = useBlockDisplayTitle( clientId );
const truncatedTitle = truncate( title, maxLength );
```

Reply:
> A cursory grep through the block.json files (there might be a few e2e fixtures in there) shows that there's not much over 28 chars in length, so we could probably get away with it.

Thank you for checking that in WordPress core. It looks like we can postpone the decision 👍🏻 

It's not that I want to push for truncating block titles, but I had this thought that I wanted to share here. When you translate titles they might be longer. As far as I remember the German language has this tendency to be much longer than what you see in English.

Reply:
> As far as I remember the German language has this tendency to be much longer than what you see in English.

That is an excellent point. 

We might one day have a _Rindfleischetikettierungsüberwachungsaufgabenübertragungsgesetz_ Block and be in real trouble. 😄 

> If the truncation is applied to both the block title and the other labels, that's probably the point at which it doesn't need to be part of the hook.

I think it's probably worth doing to make things work more logically for our future selves, thanks for the example too. I'll make a note and throw up a PR next week.

Reply:
> Rindfleischetikettierungsüberwachungsaufgabenübertragungsgesetz

I see it's a real thing 🙃

Reply:
> I'll make a note and throw up a PR next week

https://github.com/WordPress/gutenberg/pull/39416

> that's probably the point at which it doesn't need to be part of the hook.

I kept it in the hook for now to keep changes small as `<BlockTitle />` (and therefore the legacy `35` char limit) is used very liberally.